hacep/src/main/java/org/kie/hacep/consumer/CommandHandler.java

```diff
@@ -84,20 +89,39 @@ public void visit(HaltCommand command) {
     @Override
     public void visit(InsertCommand command) {
         RemoteFactHandle remoteFH = command.getFactHandle();
-        FactHandle fh = kieSessionContext.getKieSession().getEntryPoint(command.getEntryPoint()).insert(remoteFH.getObject());
-        kieSessionContext.getFhManager().registerHandle(remoteFH,
-                                                        fh);
-        if (firingUntilHalt) {
-            kieSessionContext.getKieSession().fireAllRules();
-        }
+        FactHandle fh = internalInsert( command, remoteFH.getObject() );
+        kieSessionContext.getFhManager().registerHandle(remoteFH, fh);
     }
 
     @Override
     public void visit(EventInsertCommand command) {
-        FactHandle fh = kieSessionContext.getKieSession().getEntryPoint(command.getEntryPoint()).insert(command.getObject());
-        if (firingUntilHalt) {
+        internalInsert( command, command.getObject() );
+    }
+
+    private FactHandle internalInsert( WorkingMemoryActionCommand command, Object obj ) {
+        FactHandle fh = isEvent(obj) ? insertEvent( command, obj ) : insertFact( command, obj );
+        if ( firingUntilHalt ) {
             kieSessionContext.getKieSession().fireAllRules();
         }
+        return fh;
+    }
+
+    private FactHandle insertEvent( WorkingMemoryActionCommand command, Object obj ) {
+        FactHandle fh;
+        if ( hasTimestamp( obj ) ) {
+            fh = insertFact( command, obj );
+            kieSessionContext.setClockAt( (( EventFactHandle ) fh).getStartTimestamp() );
+        } else {
+            // if the event doesn't have an its own timestamp, it has to use the command's one and then
+            // advance the pseudo clock to the command timestamp before inserting the event
+            kieSessionContext.setClockAt( command.getTimestamp() );
```

Comment:
What about a `log.info` or similar with UUID of the message to inform user that message timestamp has been used?

Reply:
If I add a log statement here this will overwhelm user's log with an info statement for each and every event instance without a timestamp. Is this what we want?

Reply:
Yes in general considering that this is class dependent I would like to have a log everytime (maybe `debug` level?) because this could generate really strange errors